cgrok/tree.c

```
@@ -0,0 +1,275 @@
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  } else {
+    free(key);
+    free(val); 
+  }
+  return inserted;
+}
+
+// Get the value for the given key, or NULL if the key is not in the tree
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  }
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  free(key);
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+// Remove all elements from the tree
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
+TCLIST *tclistnew(void) {
+  TCLIST *list = malloc(sizeof(TCLIST));
+  if (!list) {
+    return NULL;
+  }
+  list->len = 0;
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  if (!newElem) {
+    return NULL;
+  }
+  newElem->elem = 0;
+  list->head = newElem;
+  return list;
+}
+
+// Get the number of elements in the list
+int tclistnum(const TCLIST *list) {
+  return list->len;
+}
+
+// Null-terminate the value to be inserted, as TokyoCabinet does
+void *tclistpack(void *buf, uint32_t size) {
+  void *elem = malloc(size+1);
+  if (!elem) {
+    return NULL;
+  }
+  memcpy(elem, buf, size);
+  ((char *)elem)[size] = 0;
+  return elem;
+}
+
+// Append a new element at the end of the list
+void tclistpush(TCLIST *list, const void *ptr, int size) {
+  TCLISTELEM *head = list->head;
+  for (int i=0; i < list->len; i++) {
+    head = head->next;
+  }
+  TCLISTELEM *newElem = malloc(sizeof(TCLISTELEM));
+  newElem->elem = tclistpack(ptr, size);
+  if (!newElem->elem) {
+    fprintf(stderr, "Failed to malloc list element\n");
+    return;   
```

Do we need to fail fast here? There doesn't' seem to be an error code returned from this function?